client/components/tooltip/tooltip-base.js

```diff
@@ -0,0 +1,210 @@
+/**
+ * External dependencies
+ */
+import React, { useEffect, useRef, useState, memo } from 'react';
+import { createPortal } from 'react-dom';
+import classNames from 'classnames';
+import { debounce } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import './styles.scss';
+
+const rootElement =
+	document.getElementById( 'wpbody-content' ) || document.body;
+
+const isEventTriggeredWithin = ( event, element ) =>
+	element && ( element === event.target || element.contains( event.target ) );
+
+const useHideDelay = (
+	isVisibleProp,
+	{ hideDelay = 1000, triggerRef, tooltipRef, onClose }
+) => {
+	const [ isVisible, setIsVisible ] = useState( isVisibleProp );
+	// not using state for this, we don't need to cause a re-render
+	const hasMountedRef = useRef( false );
+
+	// hide delay
+	useEffect( () => {
+		let timer = null;
+
+		if ( ! hasMountedRef.current ) {
+			hasMountedRef.current = true;
+			return;
+		}
+
+		// element is marked as visible, no need to hide it
+		if ( isVisibleProp ) {
+			rootElement.dispatchEvent( new Event( 'wcpay-tooltip-open' ) );
+			setIsVisible( true );
+			return;
+		}
+
```

Comment:
This is probably not an issue because likely no one is going to be providing a custom `onClose` handler (you'd have to use `TooltipBase` directly) but this function could exit prematurely here if `isVisible` is true, otherwise the timer will be set and will fire `onClose()` even when the tooltip is already invisible.

This could happen if you pass a different `hideDelay` or `onClose` handler to `useHideDelay` even if the visibility doesn't change.

Reply:
Sorry, reading this now I realize this is not very well written.

What I had in mind was this: if the execution reaches the place where this comment is posted, it will set a timer even if the visibility hasn't changed due to the same problem as in https://github.com/Automattic/woocommerce-payments/pull/2484#discussion_r673155761.

This is resolved for `onHide` by your update to use `useRef` but not for `hideDelayMs`. I don't think this will be ran into often but if you'd like to include it, this should do the job:

```suggestion

		if ( ! isVisible ) {
			return;
		}
```

Reply:
Thanks for clarifying! Done 👍